src/trie.js

```diff
@@ -1,102 +1,518 @@
+  return {pattern, scopes, matched, paramed};
+});
+
+/**
+ * Given a sorted list of scopes that contains PARAM, replace PARAM with param
+ * and return the sorted result.
+ */
+const withParam = (scopes, param) => {
+  const pattern = param.endsWith('*') ? PARAM_TO_END : PARAM;
+  scopes = scopes.map(s => s.replace(pattern, param));
+  return ScopeSetBuilder.normalizeScopeSet(scopes);
+};
+
+/** Node in a trie */
+class Node {
+  constructor(end = [], enter = [], paramed = [], kleeneOnly = []) {
+    this.end = end;               // scopes given if input ends at this node
+    this.enter = enter;           // scopes given if input traverses through this node
+    this.paramed = paramed;       // scopes given if input traverses through this node with parameter
+    this.kleeneOnly = kleeneOnly; // scopes given in addition to end, enter, paramed if input ends with kleene
+    this._kleeneCache = null;     // scopes given if input ends with kleene at this node (lazily computed)
+    this.children = new Map();    // map from next character to child node
+  }
+
+  dump() {
+    return {
+      end: this.end,
+      enter: ScopeSetBuilder.normalizeScopeSet([...this.enter, ...withParam(this.paramed, '<..>')]),
+      kleene: this.kleeneOnly,
+      children: Array.from(this.children).reduce((result, [character, child]) => {
+        result[character] = child.dump();
+        return result;
+      }, {}),
+    };
+  }
+
+  /** Compute kleene, that is the result of reaching this node when next and final character is kleene */
+  get kleene() {
+    if (this._kleeneCache === null) {
+      const result = new ScopeSetBuilder({optionallyClone: true});
+      result.add(this.end);
+      result.add(this.enter);
+      result.add(withParam(this.paramed, '*'));
+      result.add(this.kleeneOnly);
+      for (const child of this.children.values()) {
+        result.add(child.kleene);
+      }
+      // If there is exactly one child and no end or scopes at this node,
+      // this will reuse the array from the child
+      this._kleeneCache = result.scopes();
+    }
+    return this._kleeneCache;
+  }
+
+  /**
+   * Merge node into this node after prefix, this method is the only permissible
+   * way to modify a trie, and may only be called on the root of a trie.
+   *
+   * returns true, if this node or something in its sub-tree was modified.
+   */
+  merge(prefix, node) {
+    let modified; // track if we modify this node
+    if (prefix !== '') {
+      // If there is a prefix, we find the child for the first character c of
+      // the prefix (or create such a child), and merge node into that...
+      let child = this.children.get(prefix[0]);
+      let isNew = !child;
+      if (isNew) {
+        child = new Node();
+      }
+      modified = child.merge(prefix.slice(1), node);
+      // We only save a new node if merging into it modified it, this prevents
+      // us from poluting the trie with empty nodes.
+      if (modified && isNew) {
```

Comment:
when would we have isNew and not modified?

Reply:
If we are merging in an empty node...

`isNew` is true when the child doesn't exist... `modified` is also true if merging the (new child) with the target is modifying the child... 

If we have `isNew == true && modified == false` then we've just merged an empty node into a new empty child node. There is no reason to add that child node, since we can just immediately delete it (it's empty).

I'm not 100% sure this happens. I'm pretty sure it happens a lot in `withSuffix` that we produce empty nodes... I could be wrong, because now most of them probably gets some value in `kleeneOnly`.

Reply:
I think of it this way: the optimization is safe; if I always apply those then it won't matter if I do simple things else where.

Example: `withPrefix(node, 'no-matches')` will return an empty node.

Instead of testing whether nodes are empty, or if they have sub-trees that are empty, I just always merge them and then this neat optimization ensures that empty children disappear :)